### PR TITLE
Feature/kafka

### PR DESCRIPTION
**Pull request by cichan02**

### CODE REVIEW

File: src/main/java/by/piskunou/solvdlaba/service/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -1,30 +1,47 @@
...
+import reactor.core.publisher.Mono;
+import reactor.kafka.sender.KafkaSender;
+import reactor.kafka.sender.SenderRecord;
 
 @Service
 @RequiredArgsConstructor
 public class UserServiceImpl implements UserService {
 
-    //todo: rewrite
+    private final RestTemplate restTemplate;
```
Comment (h1alexbel) [inline src/main/java/by/piskunou/solvdlaba/service/impl/UserServiceImpl.java:17]: why it's a rest template? We are using project reactor, so use webclient instead

File: src/main/java/by/piskunou/solvdlaba/service/impl/EmailServiceImpl.java [hunk 2]
```diff
@@ -1,18 +1,33 @@
...
 
+    @Override
+    public Mono<Void> sendMessage(SendEmailEvent sendEmailEvent) {
+        sender.send(
+                Mono.just(
+                        SenderRecord.create(
+                                "sendEmail",
+                                0,
+                                System.currentTimeMillis(),
+                                "key",
```
Comment (h1alexbel) [inline src/main/java/by/piskunou/solvdlaba/service/impl/EmailServiceImpl.java:25]: key shouldn't be hardcoded, consider object id or something else

File: src/main/java/by/piskunou/solvdlaba/service/JwtService.java [hunk 3]
```diff
@@ -2,21 +2,22 @@
 
 import by.piskunou.solvdlaba.domain.UserDetailsImpl;
 import org.springframework.security.core.userdetails.UserDetails;
+import reactor.core.publisher.Mono;
 
 public interface JwtService {
 
-    String extractUsername(String jwt);
+    Mono<String> extractUsername(String jwt);
```
Comment (h1alexbel) [inline src/main/java/by/piskunou/solvdlaba/service/JwtService.java:9]: I see no reason to use `Mono` or other publishers here. It's better to use Strings and Boolean, like it was before

**Review by h1alexbel — CHANGES_REQUESTED** [review]

File: src/main/java/by/piskunou/solvdlaba/service/impl/EmailServiceImpl.java (hunk 2 above)
Comment (cichan02) [reply src/main/java/by/piskunou/solvdlaba/service/impl/EmailServiceImpl.java]: add uuid to events

File: src/main/java/by/piskunou/solvdlaba/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -14,18 +14,30 @@
...
     @Override
     public User findByUsername(String username) {
-        return restTemplate.getForObject(baseUrl + "/username?" + username, User.class);
+        return webClientBuilder.baseUrl(baseUrl)
+                .build()
+                .get()
+                .uri("/username?username=" + username)
+                .retrieve()
+                .bodyToMono(User.class)
+                .block();
```
Comment (h1alexbel) [inline src/main/java/by/piskunou/solvdlaba/service/impl/UserServiceImpl.java:29]: don't block publisher

File: src/main/java/by/piskunou/solvdlaba/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -14,18 +14,30 @@
...
     @Override
     public User findByEmail(String email) {
-        return restTemplate.getForObject(baseUrl + "/email?" + email, User.class);
+        return webClientBuilder.baseUrl(baseUrl)
+                .build()
+                .get()
+                .uri("/email?email=" + email)
+                .retrieve()
+                .bodyToMono(User.class)
+                .block();
```
Comment (h1alexbel) [inline src/main/java/by/piskunou/solvdlaba/service/impl/UserServiceImpl.java:40]: don't block publisher

**Review by h1alexbel — CHANGES_REQUESTED** [review]

File: src/main/java/by/piskunou/solvdlaba/config/XMLConfig.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+@Configuration
+@Slf4j
+public class XMLConfig {
+
+    @Bean
+    public XML producerXml() throws FileNotFoundException {
+        File file = new File("/kafka/producer.xml");
+        log.info( Boolean.toString(file.exists()) );
+        return new XMLDocument(new File("/Users/solvd/IdeaProjects/micro-airport/authentication-service/src/main/resources/kafka/producer.xml"));
```
[the remark this answers is not in the document]
Comment (cichan02) [reply src/main/java/by/piskunou/solvdlaba/config/XMLConfig.java:20]: Yes, i know.
But none of that works:
/kafka/producer.xml
kafka/producer.xml
/src/main/resources/kafka/producer.xml
src/main/resources/kafka/producer.xml

[the remark this answers is not in the document]
Comment (h1alexbel) [reply src/main/java/by/piskunou/solvdlaba/config/XMLConfig.java:20]: @cichan02 fix it please

File: src/main/java/by/piskunou/solvdlaba/service/impl/JwtServiceImpl.java [hunk 7]
```diff
@@ -29,16 +28,16 @@ public Mono<String> extractUsername(String jwt) {
...
             return Mono.just( decodedJwt.getClaim("username").asString() );
         } catch (JWTDecodeException e) {
-            return Mono.error(new NoSuchClaimException("There's no username claim in jwt-token"));
+            return Mono.error(e);
         }
     }
 
     @Override
-    public Mono<String> generateAccessToken(UserDetailsImpl userDetails) {
+    public Mono<String> generateAccessToken(User user) {
```
[the remark this answers is not in the document]
Comment (cichan02) [reply src/main/java/by/piskunou/solvdlaba/service/impl/JwtServiceImpl.java:36]: Even if I use this in a controller?

@PostMapping("/generate-token/access")
@Operation(summary = "Generate access token")
@Parameter(name = "userDetails", description = "User's details for generating access token")
public Mono<String> generateAccessToken(@RequestBody User user) {
     return jwtService.generateAccessToken(user);
}

[the remark this answers is not in the document]
Comment (h1alexbel) [reply src/main/java/by/piskunou/solvdlaba/service/impl/JwtServiceImpl.java:36]: @cichan02 yeas, in this case, you just generating a simple string. why do you need Mono?

File: pom.xml [hunk 8]
```diff
@@ -199,7 +185,6 @@
             <version>${xml.version}</version>
         </dependency>
 
-        <!-- Kafka -->
         <!-- https://mvnrepository.com/artifact/org.springframework.kafka/spring-kafka -->
```
Comment (h1alexbel) [inline pom.xml:188]: please remove comments

File: pom.xml [hunk 9]
```diff
@@ -151,7 +140,6 @@
             <optional>true</optional>
         </dependency>
 
-        <!-- Mapper -->
         <!-- https://mvnrepository.com/artifact/org.mapstruct/mapstruct -->
```
Comment (h1alexbel) [inline pom.xml:143]: remove comments

**Review by h1alexbel — CHANGES_REQUESTED** [review]
@cichan02 please remove comments

**Pull request merged**